File: frontend/src/core/context.ts
export class AppContext {
  state: Record<string, unknown>
  private _serviceFactories: Record<string, (ctx: AppContext) => unknown>
  private _controllerFactories: Record<string, (ctx: AppContext) => unknown>
  private _services: Map<string, unknown>
  private _controllers: Map<string, unknown>

  constructor({
    state,
    serviceFactories,
    controllerFactories,
  }: {
    state?: Record<string, unknown>
    serviceFactories: Record<string, (ctx: AppContext) => unknown>
    controllerFactories: Record<string, (ctx: AppContext) => unknown>
  }) {
    this.state = state || {}
    this._serviceFactories = serviceFactories
    this._controllerFactories = controllerFactories
    this._services = new Map()
    this._controllers = new Map()
  }

  service(id: string): unknown {
    if (this._services.has(id)) {
      return this._services.get(id)
    }
    const factory = this._serviceFactories[id]
    if (!factory) {
      throw new Error(`Unknown service_id: ${id}`)
    }
    const service = factory(this)
    this._services.set(id, service)
    return service
  }

  controller(id: string): unknown {
    if (this._controllers.has(id)) {
      return this._controllers.get(id)
    }
    const factory = this._controllerFactories[id]
    if (!factory) {
      throw new Error(`Unknown controller_id: ${id}`)
    }
    const controller = factory(this)
    this._controllers.set(id, controller)
    return controller
  }

  reset(): void {
    this._services.clear()
    this._controllers.clear()
  }
}

File: witterungsstation_py/core/context.py
from __future__ import annotations

from typing import Any, Callable


class AppContext:
    """Small runtime context with lazy service/controller registries."""

    def __init__(
        self,
        *,
        state: dict[str, Any] | None = None,
        service_factories: dict[str, Callable[["AppContext"], Any]] | None = None,
        controller_factories: dict[str, Callable[["AppContext"], Any]] | None = None,
    ) -> None:
        self.state = state or {}
        self._service_factories = service_factories or {}
        self._controller_factories = controller_factories or {}
        self._services: dict[str, Any] = {}
        self._controllers: dict[str, Any] = {}

    def service(self, key: str) -> Any:
        name = str(key or "").strip()
        if name in self._services:
            return self._services[name]

        factory = self._service_factories.get(name)
        if factory is None:
            raise KeyError(f"Unknown service: {name}")

        instance = factory(self)
        self._services[name] = instance
        return instance

    def controller(self, key: str) -> Any:
        name = str(key or "").strip()
        if name in self._controllers:
            return self._controllers[name]

        factory = self._controller_factories.get(name)
        if factory is None:
            raise KeyError(f"Unknown controller: {name}")

        instance = factory(self)
        self._controllers[name] = instance
        return instance
